File: mergesort.py
#!/usr/bin/env python3

def merge(a , L , mid , R):
  n1 = mid - L + 1
  n2 = R - mid
  left = [0]*n1
  right = [0]*n2
  for i in range(0 , n1):
    left[i] = a[i + L]
  for i in range(0 , n2):
    right[i] = a[i + mid + 1]

  i =  0 
  j = 0 
  k = L
  while(i < n1 and j < n2):
    if(left[i] <= right[j]):
      a[k] = left[i]
      k += 1
      i += 1
    else:
      a[k] = right[j]
      k += 1
      j += 1

  while(i < n1):
    a[k] = left[i]
    k += 1
    i += 1

  while(j < n2):
    a[k] = right[j]
    k += 1
    j += 1


def mergesort(a , L , R):
  if(L < R):
    mid = ( L + R) >> 1;
    mergesort(a , L , mid)
    mergesort(a , mid + 1 , R);
    merge(a , L , mid , R)
     
if __name__ == "__main__":
  tc = int(input())
  while(tc):
    n =  int(input())
    a = (list(map(int,input().split())))
    mergesort(a , 0 , n - 1)
    for i in a:
      print(i  , end = " ")
    print()
    tc -= 1

File: languages/mergesort.swift
func merge(left: [Int],right: [Int],  _ low: Int, _ mid: Int,_  high: Int) -> [Int]{
  for i in 0...(n1 - 1){
    left.append(array[i + low])
  }
  for i in 0...(n2 - 1){
    right.append(array[i + mid + 1])
  }
  var i = 0
  var j = 0
  var res = [Int]()
  while (i < n1 && j < n2){
    if( left[i] <= right[j]){
      res.append(left[i])
      i += 1
    }
    else{
      res.append(right[j])
      j += 1
    }
  }
  while(i < n1){
    res.append(left[i])
    i += 1
  }
  while(j < n2){
    res.append(right[j])
    j += 1
  }
  var k = low
  for el in res{
    array[k] = el
    k += 1
  }
}


func mergesort(array: [Int], _ low: Int, _ high: Int) -> [Int]{
  if(array.count == 1){
    return array
  }
  if(array.count == 2){
    return merge([a[0]], [a[1]])
  }
  if (low < high){
    var mid = (low + high) >> 1;
    let a1 =mergesort(array, low, mid)
    let a2 = mergesort(array,mid + 1,high)
    return merge(array, low, mid,  high)
  }
}

var a = [8, 9, 1, 0];
print(mergesort(a , 0, 3))
